### PR TITLE
FIX time overhead when fitting with ``Lasso`` estimator

### LINKED ISSUE #128: Considerable overhead when fitting with ``Lasso``
## Description

We have a considerable overhead when fitting Lasso Estimator as shown in the screenshot below

</img>

To reproduce go to [benchopt Lasso benchmark repo](https://github.com/benchopt/benchmark_lasso)

## Investigation

After investigating, this overhead is because of the computation of the ``global_lipschitz`` 

https://github.com/scikit-learn-contrib/skglm/blob/cca6d48b57bf2e9e69fb4dac042246f66aff1b21/skglm/datafits/single_task.py#L52-L54

that we introduced after adding the FISTA solver https://github.com/scikit-learn-contrib/skglm/pull/91.

## Potential fix

The ``global_lipschitz`` is only relevant for the FISTA solver. Hence, it should be computed only in this case.

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
This provides a quick fix of the time overhead that one encounters when fitting a Lasso estimator.
A more robust fix will come in another PR.

close #128

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
[link to bench results](https://badr-moufad.github.io/share-benchmarks/investigation_andcd/benchmark_lasso_benchopt_run_2022-11-30_12h05m23.html)

Back to shape :muscle:!

File: skglm/solvers/fista.py [hunk 1]
```diff
@@ -39,23 +43,39 @@ def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
-            # TODO: OR line search
-            raise Exception("Line search is not yet implemented for FISTA solver.")
+            lipschitz = norm(X, ord=2)**2 / n_samples
 
         for n_iter in range(self.max_iter):
             t_old = t_new
             t_new = (1 + np.sqrt(1 + 4 * t_old ** 2)) / 2
             w_old = w.copy()
-            if issparse(X):
+            if is_X_sparse:
```
Comment (mathurinm) [inline skglm/solvers/fista.py:78]: nitpick: `if X_is_sparse` reads better IMO

File: skglm/tests/test_fista.py [hunk 2]
```diff
@@ -7,9 +7,9 @@
 import scipy.sparse.linalg
 from scipy.sparse import csc_matrix, issparse
 
-from skglm.penalties import L1, IndicatorBox
+from skglm.penalties import L1
 from skglm.solvers import FISTA, AndersonCD
-from skglm.datafits import Quadratic, Logistic, QuadraticSVC
```
Comment (mathurinm) [inline skglm/tests/test_fista.py:12]: why do we remove these?

**Review by mathurinm — CHANGES_REQUESTED** [review]
Instead of the current approach I'd define datafit.compute_lipschitz() and datafit.compute_lipschitz_sparse() in order to not delete existing computations in datafits

This method would compute the LC and set it as an attribute of the df, so we'd only change two lines in fista(), and remove the ValueError etc.

File: skglm/datafits/single_task.py [hunk 3]
```diff
@@ -328,24 +340,26 @@ def params_to_dict(self):
...
             for idx in range(X_indptr[j], X_indptr[j + 1]):
                 nrm2 += X_data[idx] ** 2
             self.lipschitz[j] = nrm2 / len(y)
 
+    def init_global_lipschitz(self, X, y):
+        self.global_lipschitz = norm(X, ord=2) ** 2 / len(y)
+
+    def init_global_lipschitz_sparse(self, X_data, X_indptr, X_indices, y):
+        self.global_lipschitz = spectral_norm(X_data, X_indptr, X_indices, len(y)) ** 2
+        self.global_lipschitz /= len(y)
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:361]:
```suggestion
        self.global_lipschitz = spectral_norm(
            X_data, X_indptr, X_indices, len(y)) ** 2 / len(y)
```

File: skglm/datafits/single_task.py [hunk 4]
```diff
@@ -155,19 +160,22 @@ def raw_hessian(self, y, Xw):
...
         for j in range(n_features):
             Xj = X_data[X_indptr[j]:X_indptr[j+1]]
             self.lipschitz[j] = (Xj ** 2).sum() / (len(y) * 4)
 
+    def init_global_lipschitz(self, X, y):
+        self.global_lipschitz = norm(X, ord=2) ** 2 / (4 * len(y))
+
+    def init_global_lipschitz_sparse(self, X_data, X_indptr, X_indices, y):
+        self.global_lipschitz = spectral_norm(X_data, X_indptr, X_indices, len(y)) ** 2
+        self.global_lipschitz /= 4 * len(y)
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:177]:
```suggestion
        self.global_lipschitz = spectral_norm(
            X_data, X_indptr, X_indices, len(y)) ** 2 / 4 * len(y)
```

File: skglm/solvers/fista.py [hunk 5]
```diff
@@ -39,23 +39,27 @@ def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
-        else:
-            # TODO: OR line search
-            raise Exception("Line search is not yet implemented for FISTA solver.")
+        try:
+            if X_is_sparse:
+                datafit.init_global_lipschitz_sparse(X.data, X.indptr, X.indices, y)
+            else:
+                datafit.init_global_lipschitz(X, y)
+        except AttributeError:
+            raise Exception("Datafit is not compatible with FISTA solver.")
```
Comment (mathurinm) [inline skglm/solvers/fista.py:55]: + must implement init_global_lipschitz   (to give actionable feedback to the user)

**Review by mathurinm — COMMENTED** [review]
few cosmetic changes, apart from that looks like an efficient quick fix @Badr-MOUFAD

**Review by mathurinm — APPROVED** [review]

**Comment by mathurinm** [comment]
Thanks @Badr-MOUFAD 

@PABannier this may explain the bad perf you had on Finance recently

**Pull request merged**